inspectit-oce-core/src/main/java/rocks/inspectit/oce/core/instrumentation/InstrumentationManager.java

```diff
@@ -0,0 +1,209 @@
+     * This required because in case of a config change we need to reevalaute the intrumentation for each class.
+     */
+    private Set<Class<?>> loadedClasses = Collections.newSetFromMap(
+            CacheBuilder.newBuilder().weakKeys().<Class<?>, Boolean>build().asMap());
+
+    /**
+     * The set of classes which might need instrumentation updates.
+     * This service works through this set in batches.
+     */
+    private Set<Class<?>> pendingClasses = Collections.newSetFromMap(
+            CacheBuilder.newBuilder().weakKeys().<Class<?>, Boolean>build().asMap());
+
+    private BatchJobExecutorService.BatchJob<BatchSize> classInstrumentationJob;
+
+    @Value
+    private static class BatchSize {
+        private int maxClassesToCheck;
+        private int maxClassesToRetransform;
+    }
+
+
+    @PostConstruct
+    private void init() {
+        InternalSettings conf = env.getCurrentConfig().getInstrumentation().getInternal();
+        val batchSizes = new BatchSize(conf.getClassConfigurationCheckBatchSize(), conf.getClassRetransformBatchSize());
+        Duration delay = conf.getInterBatchDelay();
+
+        classInstrumentationJob = executor.startJob(this::checkClassesForConfigurationUpdates, batchSizes, delay, delay);
+    }
+
+    @PreDestroy
+    private void destroy() {
+        classInstrumentationJob.cancel();
+    }
+
+    @Override
+    public void newClassesDiscovered(Set<Class<?>> newClasses) {
+        loadedClasses.addAll(newClasses);
+        pendingClasses.addAll(newClasses);
+    }
+
+    @EventListener
+    private void classInstrumented(ClassInstrumentedEvent event) {
+        ClassInstrumentationConfiguration config = event.getAppliedConfiguration();
+        TypeDescription type = event.getInstrumentedClassDescription();
+        Class<?> clazz = event.getInstrumentedClass();
+
+        if (ClassInstrumentationConfiguration.NO_INSTRUMENTATION.isSameAs(type, config)) {
+            activeInstrumentations.remove(clazz);
+        } else {
+            activeInstrumentations.put(clazz, config);
+        }
+
+        //there might be race conditions when a class is being retransformed
+        //and we considered it up-to-date at the same time
+        //for this reason we have to recheck every class after it has been instrumented
+        pendingClasses.add(clazz);
+    }
+
+    @EventListener
+    private void configEventListener(InspectitConfigChangedEvent ev) {
+
+        InternalSettings newInternal = ev.getNewConfig().getInstrumentation().getInternal();
+        val batchSizes = new BatchSize(newInternal.getClassConfigurationCheckBatchSize(), newInternal.getClassRetransformBatchSize());
+        classInstrumentationJob.setBatchSizes(batchSizes);
+        classInstrumentationJob.setInterBatchDelay(newInternal.getInterBatchDelay());
+    }
+
+    @EventListener
+    private void instrumentationConfigEventListener(InstrumentationConfigurationChangedEvent ev) {
+        pendingClasses.addAll(loadedClasses);
```

Comment:
Now we are holding twice the complete loaded classes in a set, here and in the discovery service.. would it make sense that we move this event listener to the discovery service and report via the listener? this way we would keep loaded classes only in the discovery service.. maybe we should then rename the `newClassesDiscovered(Set<Class<?>> newClasses)` to something more meaningful?

Reply:
let's just get them all from the `Instrumentation`